### PR TITLE
implement error write protection circuit

### PR DESCRIPTION
**Pull request by DreamWuGit**
- [x]  fix get_step_error parsing call with value #1010 .
  - [x]  implement constraints: write protection error only happens in some opcodes which modify states(e.g. create, sstore, log*, call etc.) for call, need to special check value is not zero. and this error only happens in internal call since requires at least one staticcall in the before.
  - [x] add tests for sstore, call. will add more in the future. 
  
Note:  since it relies on PR #1010 , so will make it ready for review after that merged. and  will add spec after this PR got merged !!

SPEC in https://github.com/privacy-scaling-explorations/zkevm-specs/pull/365

### CODE REVIEW

**Comment by aguzmant103** [comment]
Hey @DreamWuGit can you help us get a 2nd reviewer? Thanks a lot!

**Comment by DreamWuGit** [comment]
> Hey @DreamWuGit can you help us get a 2nd reviewer? Thanks a lot!

@aguzmant103  sorry, just see it, you mean second one from scroll  or anyone?

**Comment by DreamWuGit** [comment]
@icemelon  or @lispc  will take second review from scroll side.

**Comment by DreamWuGit** [comment]
> @DreamWuGit do we have the specs written for this implementation? Usually goes in https://github.com/privacy-scaling-explorations/zkevm-specs/tree/master/specs/error_state

will do it after circuit PR got approved to avoid spec re-factoring work after real circuit implementation done.

File: bus-mapping/src/evm/opcodes/error_write_protection.rs
Comment (DreamWuGit) [reply bus-mapping/src/evm/opcodes/error_write_protection.rs:42]: I remember state circuit will check stack ops with stack pointer increase/decrease by one . so not skip.

File: zkevm-circuits/src/evm_circuit/execution/error_write_protection.rs
Comment (DreamWuGit) [reply zkevm-circuits/src/evm_circuit/execution/error_write_protection.rs:45]: good to me, thanks, will update it.

Comment (DreamWuGit) [reply zkevm-circuits/src/evm_circuit/execution/error_write_protection.rs]: updated in [dd3dc7b](https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1001/commits/dd3dc7bf7e4146ccda50582cb203a1a9b3fa63d5) thanks for catching!

**Comment by aguzmant103** [comment]
@DreamWuGit heya! Just following up with you this week regarding the md spec

**Comment by DreamWuGit** [comment]
> @DreamWuGit heya! Just following up with you this week regarding the md spec

@aguzmant103 @adria0 will add md spec & follow: specs > circuit > python.  thanks!

**Comment by DreamWuGit** [comment]
@adria0 @aguzmant103 would you like to have a review ?

**Comment by DreamWuGit** [comment]
> @DreamWuGit, waiting for [privacy-scaling-explorations/zkevm-specs#365 (comment)](https://github.com/privacy-scaling-explorations/zkevm-specs/pull/365#issuecomment-1427528435) also my comment on transition to EndTx when is_root.

it couldn't be root call because it will have prior static call , only in static call then execute modifying op code will meet write protection error.  if in root call, such error will never happen

File: zkevm-circuits/src/evm_circuit/execution/error_write_protection.rs
Comment (DreamWuGit) [reply zkevm-circuits/src/evm_circuit/execution/error_write_protection.rs:63]: good note .

File: zkevm-circuits/src/evm_circuit/execution/error_write_protection.rs
Comment (DreamWuGit) [reply zkevm-circuits/src/evm_circuit/execution/error_write_protection.rs:77]: sure

File: zkevm-circuits/src/evm_circuit/execution/error_write_protection.rs
Comment (DreamWuGit) [reply zkevm-circuits/src/evm_circuit/execution/error_write_protection.rs:299]: could be , will have a try !

Comment (DreamWuGit) [reply zkevm-circuits/src/evm_circuit/execution/error_write_protection.rs]: changed to use `STOP`.

**Comment by DreamWuGit** [comment]
> BTW, also checked that `SSTORE` in the tx init code for deployment does not triggers a WriteProtection

that is correct behavior !!

**Review by lispc — APPROVED** [review]
Lgtm But i think we d better first Merge the common error gadget pr？

**Review by silathdiir — APPROVED** [review]

**Comment by DreamWuGit** [comment]
> Lgtm But i think we d better first Merge the common error gadget pr？

I described in common error gadget pr to wait for previous(this) pr and sync then.

**Pull request merged**
